dev-docs: refactor and add information for newbies (#1912)
edgelesssys/constellation: dev-docs/workflows/build-develop-deploy.md

Reviewer: @thomasten PTAL at the OSS / enterprise licensing aspects

# Getting started locally

The following are instructions for building all components in the constellation repository, except for images. A manual on how to build images locally can be found in the [image README](/image/README.md).

Prerequisites:

* 20GB (minimum), better 40 GB disk space (required if you want to cross compile for all platforms)
* [Latest version of Go](https://go.dev/doc/install).
* [Bazelisk installed as `bazel` in your path](https://github.com/bazelbuild/bazelisk/releases).

> * [Docker](https://docs.docker.com/engine/install/). Can be installed with these commands on Ubuntu 22.04: `sudo apt update && sudo apt install docker.io`. As the build spawns docker containers your user account either needs to be in the `docker` group (Add with `sudo usermod -a -G docker $USER`) or you have to run builds with `sudo`. When using `sudo` remember that your root user might (depending on your distro and local config) not have the go binary in it's PATH. The current PATH can be forwarded to the root env with `sudo env PATH=$PATH <cmd>`.

Reviewer: Most but not all. Keeping this for now is probably the right call.

## Prequisites

### Linux

* Packages on Ubuntu:

  ```sh
  sudo apt install build-essential cmake libssl-dev pkg-config libcryptsetup12 libcryptsetup-dev
  ```

* Packages on Fedora:

  ```sh
  sudo dnf install @development-tools pkg-config cmake openssl-devel cryptsetup-libs cryptsetup-devel
  ```

> ### Mac
> 
> ```sh
> brew install bash
> ```
> 
> to fix unsupported shell options used in some build script.
> 
> To troubleshoot potential problems with bazel on ARM architecture when running it for the first time, it might help to purge and retry:
> 
> ```sh
> bazel clean --expunge
> ```

Reviewer 2: ```suggestion
### Mac

* To fix unsupported shell options used in some build script:

  ´´´sh
  brew install bash
  ´´´

* To troubleshoot potential problems with bazel on ARM architecture when running it for the first time, it might help to purge and retry:

  ´´´sh
  bazel clean --expunge
  ´´´
```

Align formatting with the prerequisites for Linux
(Ignore my change in backticks, had to updated them otherwise GitHub suggestion formatting breaks)

## Settings

### Set build version

There exists an `oss` and an `enterprise` version for licensing reasons.

`enterprise`: recommended, complete feature set but some features are not licensed under Open-Source

> `oss`: default when building from the OSS repo, but non OSS-licensed features are missing (e.g. verify boot measurements). The [official docs](https://docs.edgeless.systems/constellation/) are written to be compatible with `oss`.

Reviewer 2: ```suggestion
`oss`: default when building from the OSS repo, but non OSS-licensed features are missing (e.g. fetching expected measurements). The [official docs](https://docs.edgeless.systems/constellation/) are written to be compatible with `oss`.
```
Verifying boot measurements should always be supported, since oss/enterprise feature gates only influence the behavior of the CLI, not the whole Constellation.

Also not sure about the docs, I think its the other way around? Maybe we should discuss during standup

<!-- are we sure about doc support? -->

For convenience, we can set `enterprise` as the default build version:

```

> cat .bazeloverwriterc
> build --cli_edition=enterprise

Reviewer 2: ```suggestion
echo "build --cli_edition=enterprise" >> .bazeloverwriterc
```
We could also add a sample `.bazeloverwriterc` which includes all the relevant overwrite flags to `bazel.md`

```

### Authenticate with the GitHub registry

Get a GitHub token with access to the registry and authenticate through `docker` (see [here](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry)).

Optionally, you can customize the default registry target:

```sh

> cat .bazeloverwriterc
> build --container_prefix=ghcr.io/USERNAME/constellation

Reviewer 2: ```suggestion
echo "build --container_prefix=ghcr.io/USERNAME/constellation" >> .bazeloverwriterc
```

build --cli_edition=enterprise
```

### Authenticate with the GitHub registry

Get a GitHub token with access to the registry and authenticate through `docker` (see [here](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry)).

Optionally, you can customize the default registry target:

```sh
cat .bazeloverwriterc
build --container_prefix=ghcr.io/USERNAME/constellation
```

### Authenticate with your cloud provider

To provision the constellation cluster on the provider infrastructure, please authenticate with respective provider.

E.g. AWS:

```sh
aws login
```

For more details, see [here](https://docs.edgeless.systems/constellation/getting-started/install#set-up-cloud-credentials).

## Build

Build the binaries always in a separate directory:

```sh
mkdir build && cd build
# build required binaries for a dev build
# and symlink them into the current directory
# also push the built container images
# After the first run, set the pushed imaged to public.
bazel run //:devbuild --cli_edition=oss --container_prefix=ghcr.io/USERNAME/constellation
./constellation ...
```

### Specialized Bazel build targets

> see [bazel](./bazel.md#build)

Reviewer 2: ```suggestion
See [bazel](./bazel.md#build).
```

## Develop & Contribute

### VS Code setup

We recommend to set up your IDE to conform with our conventions (see [here](../conventions.md/#recommended-vs-code-settings)). You might also consider to set up Bazel in the IDE (see [here](./bazel.md#vs-code-integration)).

### Testing

See [here](./testing.md)

### Pre-Opening a PR

> 
> Before opening a PR, please run the tests and
> 
> ```sh
> bazel run //:generate && bazel run //:tidy
> bazel run //:check
> ```

Reviewer 2: ```suggestion

Before opening a PR, please run the tests and regenerate all files:

´´´sh
bazel run //:generate && bazel run //:tidy
bazel run //:check
bazel test //...
´´´
```

We can easily include the bazel test target here.
(Also please ignore my backticks change again)

Please note that some errors thrown in `check` by `golicenses_check` are ignored (for more see [golicenses.sh.in](../../bazel/ci/golicenses.sh.in)).

### PR-Guidelines

Please read the [PR conventions](https://github.com/edgelesssys/constellation/blob/main/dev-docs/conventions.md#pr-conventions).
Furthermore, the PR titles are used for the changelog, so please stick to the naming convention.

## Deploy

> :warning: Debug images are not safe to use in production environments. :warning:
The debug images will open an additional **unsecured** port (4000) which accepts any binary to be run on the target machine. **Make sure that those machines are not exposed to the internet.**

### Cloud

To familiarize yourself with debugd and learn how to deploy a cluster using it, read [this](/debugd/README.md) manual.
If you want to deploy a cluster for production, please refer to our user documentation [here](https://docs.edgeless.systems/constellation/getting-started/first-steps#create-a-cluster).

### Locally

In case you want to have quicker iteration cycles during development you might want to setup a local cluster.
You can do this by utilizing our QEMU setup.

> Instructions on how to set it up can be found in the [QEMU README](qemu.md).

Reviewer 2: ```suggestion
Instructions on how to set it up can be found in [our docs](https://docs.edgeless.systems/constellation/getting-started/first-steps-local#create-a-cluster). See the [QEMU README](qemu.md) for additional details.
```
QEMU support is now in our official docs and can be referenced here
